File: 1871-jump-game-vii/1871-jump-game-vii.cpp
class Solution {
public:
    bool canReach(string s, int minJump, int maxJump) {
        int n = size(s);
        multiset<int> ms;
        vector<int> dp(n);
        dp[0] = 1;
        for(int i = 1; i < n; i++) {
            if(i >= minJump) {
                ms.insert(dp[i - minJump]);
            }
            if(i > maxJump) {
                ms.erase(ms.find(dp[i - maxJump - 1]));
            }
            if(s[i] == '0' && !ms.empty() && *ms.rbegin() == 1) {
                dp[i] = 1;
            }
        }
        return dp[n - 1];
    }
};

File: 1871-jump-game-vii/1871-jump-game-vii.go
func canReach(s string, minJump int, maxJump int) bool {
    n := len(s)
    cnt := 0
    var dp []int
    dp = make([]int, n)
    dp[0] = 1
    for i := 1; i < n; i++ {
        if i >= minJump {
            cnt += dp[i - minJump]
        }
        if i > maxJump {
            cnt -= dp[i - maxJump - 1];
        }
        if s[i] == '0' && cnt > 0 {
            dp[i] = 1
        }
    }
    if dp[n - 1] == 1 {
        return true
    } else {
        return false
    }
}